### ISSUE TITLE
`api-supported-version` does not honor actions, only controllers

### ISSUE THREAD

**Issue by MoaidHathot** [microsoft/aspnet-api-versioning#754]
The `api-supported-version` does not honor actions, it is generated according to controller aggregation.
For example, assuming we have two versions `v2021-07-01-preview` and `v2021-08-01-preview`, and the below structure.

As you can see below, `[GET] /test/Bar` exists only in `v2021-08-01-preview`, and when I try to call it with API Version `v2021-07-01-preview` I get `UnsupportedApiVersion` as expected, but in the response headers, in `api-supported-versions` I get also `2021-07-1-preview` which does not support this action.

```
allow: GET 
 api-supported-versions: 2021-07-01-preview2021-08-01-preview 
 content-type: application/json; charset=utf-8 
 date: Thu08 Jul 2021 12:11:33 GMT 
 server: Kestrel 
```

Am I doing something wrong? Can we change this behavior?

Project structure and classes:
```
Project
|
\Controllers
    |
    \v2021_07
         |
         ./TestController.cs
    |\v2021_08
        |
        ./TestController.cs
```
  
The older controller:
```csharp
namespace ApiVersioningTest.Controllers.v2021_07
{
    [ApiController]
    [ApiVersion("2021-07-01-preview")]
    [Route("test")]
    public class TestController : ControllerBase
    {
        [HttpGet]
        [Route("Foo")]
        public IActionResult Foo()
            => Ok("This is Foo: 2021-07-01-preview");
    }
}
```

The newer version:
```csharp
namespace ApiVersioningTest.Controllers.v2021_08
{
    [ApiController]
    [ApiVersion("2021-08-01-preview")]
    [Route("test")]
    public class TestController : ControllerBase
    {
        [HttpGet]
        [Route("Foo")]
        public IActionResult Foo()
            => Ok("This is Foo: 2021-08-01-preview");

        [HttpGet]
        [Route("Bar")]
        public IActionResult Bar()
            => Ok("This is Bar: 2021-08-01-preview");
    }
}
```

```csharp
public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        public void ConfigureServices(IServiceCollection services)
        {

            services.AddMvc();

            services.AddControllers();

            services.AddApiVersioning(options => {
                options.ReportApiVersions = true;
            });

            services.AddVersionedApiExplorer();


            services.AddSwaggerGen(c =>
            {
                c.SwaggerDoc("2021-08-01-preview", new OpenApiInfo { Title = "2021-08-01-preview", Version = "2021-08-01-preview" });
                c.SwaggerDoc("2021-07-01-preview", new OpenApiInfo { Title = "2021-07-01-preview", Version = "2021-07-01-preview" });
            });
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseSwagger();
                app.UseSwaggerUI(options =>
                {
                    options.SwaggerEndpoint("/swagger/2021-08-01-preview/swagger.json", "2021-08-01-preview");
                    options.SwaggerEndpoint("/swagger/2021-07-01-preview/swagger.json", "2021-07-01-preview");
                });
            }

            app.UseHttpsRedirection();

            app.UseRouting();

            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });
        }
    }   
```

**Comment by commonsensesoftware** [comment]
Currently, this is by design. While I agree that it's not unreasonable to group by action, it's much more complicated than it appears. Grouping `test/bar` is pretty simple, but things quickly get much more complex. Consider that `values/{id}` and `values/{id:int}` are semantically equivalent (e.g. equal). They _should_ be grouped together, but it requires a much deeper evaluation of the route template.

Now consider something more elaborate such as `values/{id}/subvalues`. If grouping is purely on action, then all actions are treated as root-level APIs when they might not be. RESTful APIs are typically resource-based, but they don't have to be. _Containment_ around those resources aren't necessarily separate APIs. More importantly, there isn't a clear way to determine what a developer means by route template alone. This _could_ lead to a situation where API versions for actions are incongruent with each other unintentionally. Another issue will be how APIs are collated for OpenAPI/Swagger. Versioned APIs are bucketized by API version and then by controller/resource name.

Today, the only reliable and sensible way to group logical sets of APIs together is by controller name. I've been thinking about new ways to facilitate this so developers have more control over that grouping; for example, what if there is no controller? You still need to group, but I don't think that route template alone will ever be enough. This will mean that some new facility will need to be added so you can specify the logical name of an endpoint to decide which APIs should be grouped together.

The `api-supported-versions` and `api-deprecated-versions` was never meant to define the API versions for a specific action, although they could. Detailed API information can be comprehensive and is better supported by OpenAPI documents. Two useful things that these values can provide are:

- Which API versions are supported and deprecated?
   -  OpenAPI doesn't tell you this
   - You could implement `OPTIONS`, which would return the aggregated set to choose an OpenAPI document
- Is the API version I'm using deprecated?
   - This provides an automated communication path to clients who honor it to let them know their API will be going away at some point

That's probably not the answer you were looking or hoping for, but I hope it clears things up - if not a little.

TL;DR

A couple of side notes...

You don't have to declare actions as separate routes. In classic Web API, you had `[RoutePrefix]` and `[Route]`, where `[Route]` on the action mapped the endpoint. This is almost backward in ASP.NET Core. Adding the suffix to the HTTP method route constraint will suffice. You can also use the `[controller]` and `[action]` tokens. `[action]` uses the name of the method for the path segment. You could also use `nameof(Foo)`. This helps avoid a class of mistakes if you refactor the method name. For example:

```c#
[ApiController]
[ApiVersion("2021-08-01-preview")]
[Route("[controller]")]
public class TestController : ControllerBase
{
    // GET test/foo
    [HttpGet("[action]")]
    public IActionResult Foo()
        => Ok("This is Foo: 2021-08-01-preview");

    // GET test/bar
    [HttpGet("[action]")]
    public IActionResult Bar()
        => Ok("This is Bar: 2021-08-01-preview");
}
```

If you follow the [Swagger Sample](../blob/master/samples/aspnetcore/SwaggerSample), you'll see that you don't need to list the documents manually. You can build them with simple loops based on the collated set of API versions. This is a big part of the feature support for the API Explorer extensions so you don't have to keep changing that code when versions are added or removed. [Startup.cs](../blob/master/samples/aspnetcore/SwaggerSample/Startup.cs#L79) will show you how to map the endpoints and [ConfigureSwaggerOptions.cs](../blob/master/samples/aspnetcore/SwaggerSample/ConfigureSwaggerOptions.cs) will show you how to create new document information for each API version.

**Comment by MoaidHathot** [comment]
Thank you @commonsensesoftware for your prompt and elaborate response. I really appreciate the effort and that you shared the cause, thoughts, and design for this behavior with us 😊

Also thank you for the tips regarding the route and swagger. I will take them into account!

**Comment by Sirozha1337** [comment]
I'm running into this problem now, trying to make the client log useful info when versions are mismatched. 

Example code:
```dotnet
[ApiController]
[ApiVersion("1.0")]
[Route("[controller]")]
public class TestController : ControllerBase
{
    // GET test/foo
    [HttpGet("[action]")]
    [ApiVersion("1.0")]
    public IActionResult Foo()
        => Ok("This is Foo: 1");

    // GET test/bar
    [HttpGet("[action]")]
    [ApiVersion("2.0")]
    public IActionResult Bar()
        => Ok("This is Bar: 2");
        
    // GET test/xyz
    [HttpGet("[action]")]
    [ApiVersion("3.0")]
    public IActionResult Xyz()
        => Ok("This is Xyz: 3");
}
```

When calling the method "Xyz" with version "2", you will get "Unsupported API version" with "api-supported-versions: 1.0, 2.0, 3.0". Looking at such info in logs will seem strange, like why would the method fail, when it clearly supports the requested version?!

Is there some workaround for that, other than creating a new controller for each method?

**Comment by commonsensesoftware** [comment]
@Sirozha1337 The issue isn't so much about a new controller or action. An API is more than a single endpoint. An _API_ often has several different endpoints. The _Orders_ API, for example, might have a `GET`, `POST`, and `DELETE` method endpoint. In your example, you have defined the _Test_ API. Collation occurs at the _API_ level, which is its logical name. If you are using controllers, then it will be the controller name. There really is no other way that collation can be done.

As mentioned above, a set of API versions alone cannot fully describe what endpoint may be available for a particular version. Other tools such as OpenAPI are better suited for that. What `api-supported-versions` tells you is what versions of an API, the entire API, are supported. This may seem unintuitive when you interpolate versions on a single controller, but it becomes more apparent when you split versions into different controller types, which is the more common approach. There is a lot of fragility in trying to collate by action template, name, etc.

Exactly how a client would even get into this situation varies. This would mostly likely occur if a client bound to an older API version updates without OpenAPI and just assumes the same endpoints are supported in the next version. There isn't a clear way to indicate the list of supported endpoints so the message is a little misleading. The message should be interpreted to be read as _"The endpoint you requested has a supported version - possibly several, but not the one you asked for."_

There are likely some _workarounds_, but they are non-trivial and would be very specific to the way you implemented versioning on the server side. For example, you might collate by full route template because you know and have control over those templates. Implementing that in a generic way for every possible application is much more difficult.

If you were not aware, there is a client-side, API version-aware library to facilitate both including the API version in requests automatically and logging scenarios such as when you are using a deprecated version or when a new version becomes available. If you use the newer _Sunset Policy_ feature, a client may even know exactly when a particular version will sunset for good.

**Comment by Sirozha1337** [comment]
@commonsensesoftware thanks for the detailed answer. While I understand why it was done this way, I think it should be a customizable feature. I tried replacing the services in DI with my own implementations, but it feels hacky and doesn't seem to work correctly for my case. If you can share some of the known workarounds, that would be great. My action templates do not contain path parameters, so it should be easy to group by endpoint's path. Also my API is an internal one and I do not intend to support multiple versions for longer than a week or so, just until I update my other services, that are consuming the API.

**Comment by commonsensesoftware** [comment]
@Sirozha1337,

_"I knew you'd say that!"_ 🤣 

Changing the way things are aggregated is not trivial and is deeply involved. Save for a few basic tweaks, there aren't a lot of hooks at that level. I probably _could_ do it, but I'm not sure it's worth it. Fortunately, there should be an easier way.

You're not the first person to ask this. I baked in a change some time back that _should_ get you want you want. You likely don't need to change how aggregation is done, but rather how _reporting_ is done. Every endpoint has an `ApiVersioningMetadata` that consists of an optional, logical name, an `ApiVersionModel` for the overall API, and an `ApiVersionModel` for a specific endpoint. The `ApiVersionModel` for the endpoint is referred to as the _explicit_ model, while the overall API variant is the _implicit_ model. You can ask for either or a union of both via the `ApiVersionMapping` enumeration.

The `IReportApiVersions` service is what _reports_ API versions in responses. It now has a `Mapping` property for the `ApiVersionMapping` enumeration so that consumers know which mapping to use to generate the model before reporting it. The default service registration will include both implicit and explicit mappings. You can reconfigure and reregister the service without having to create a custom implementation like so:

```c#
services.Replace(ServiceDescriptor.Singleton<IReportApiVersions>(sp =>
  new DefaultApiVersionReporter(
    sp.GetRequiredService<ISunsetPolicyManager>(),
    mapping: ApiVersionMapping.Explicit)));
```

If your APIs are interleaved on a single controller with explicit `[ApiVersion]`s declared on your actions, then this configuration _should_ work for you. If you start trying to split things apart into different types, this will probably not work because the API versions defined by different actions will be considered _implicit_, even though they have the same route template.

Let's start with that and see if it works for you. I have some ideas about how you might break things apart without having to completely change the default aggregation, but it's going to be quite involved.

**Comment by Sirozha1337** [comment]
Thanks! Sadly, I'm still on the .NET 6 (sorry for not mentioning that earlier). I changed it like this, but it still didn't work:
```csharp
services.Replace(ServiceDescriptor.Singleton<IReportApiVersions>(sp =>
    new DefaultApiVersionReporter(
        mapping: ApiVersionMapping.Explicit)))
 ```
 It reports all the versions available in the controller as supported by the endpoint.
 
 Guess, I'll just have to live with that and do not include those versions when handling errors in client code.

**Comment by commonsensesoftware** [comment]
@Sirozha1337 Ah - yes - there is a slight difference between the `6.x` and `7.0`+ setup, but it looks like you figured it out.

Actually, I got ahead of myself. You're correct; this won't work. In fact, I'm not sure that it _can_ be made to work. The reason that it can't work (easily) is that you want to report the supported versions when an request **does not** match back to the client. That is extremely challenging. How can you report any API versions if you don't know the target endpoint? The endpoint is unmatched.

There was actually no way to initially do this at all when `6.0` was released. This was a breaking behavior change from days gone by. This happened because changes to routing itself were required to get the proper responses for `405`, `406`, and `415`; the API version has to be mapped earlier in the route tree. Then some people wanted the reporting functionality back.

I had to think long and hard about it, but I was finally able to restore parity with the `<= 5.x` behavior. I effectively rely on the routing system building up the tree mapping using similar route templates. This _feels_ brittle, but so far it seems work out, even though it's technically _best effort_. For `6.x`, you can take a peek here:

https://github.com/dotnet/aspnet-api-versioning/blob/e9242d65295f2e9d18ab324d0776c85e75321ab6/src/AspNetCore/WebApi/src/Asp.Versioning.Http/Routing/ApiVersionMatcherPolicy.cs#L120

and here:

https://github.com/dotnet/aspnet-api-versioning/blob/e9242d65295f2e9d18ab324d0776c85e75321ab6/src/AspNetCore/WebApi/src/Asp.Versioning.Http/Routing/ApiVersionMatcherPolicy.cs#L355

It would probably technically be possible to add a service that could control aggregation, but I'm still not sure it helps. Assuming you could bucketize by route template and API versions, the challenge is that you still don't have a match. You'd have to determine the _best non-match_, which seems really complex and brittle.

This is an area where I wouldn't mind improving things. If you have some ideas about how that may be possible, I'm willing to explore it a bit further. It should probably be queued up a new enhancement or feature request.

**Comment by Sirozha1337** [comment]
@commonsensesoftware ah, I see. So it routes by both the version and route at the same time? I guess, aggregation won't help there at all then... 

Thanks a lot for the pointers, though! I'll look into the code.

**Comment by commonsensesoftware** [comment]
@Sirozha1337 If you're really that brave to dive down to that level, then you will likely find the DAG generated in acceptance tests useful:

https://github.com/dotnet/aspnet-api-versioning/blob/2292fbe6a1598d944cd5cbca918cb79da7339116/src/AspNetCore/Acceptance/Asp.Versioning.Mvc.Acceptance.Tests/HttpServerFixture.cs#L38

When you run any of the acceptance tests, this link will be generated for the fixture. Fixture setup is expensive, so expect that there is probably only one DAG, even if the link is generated many times (at the test level). You can take the link generated and paste it into a browser. That will provide a visualization of the routing tree as a DAG. More than anything, it will provide a visualization as to how routing decisions are made. I found this little gem to be very useful for debugging.

**Comment by xavierjohn** [comment]
It is `ApiVersionMetadata`, not `ApiVersioningMetadata`.
